### PR TITLE
bt-149: Add socket plugin

### LINKED ISSUE #149: Add socket package
- [x] Socket package
- [x] Join chat event
- [x] Leave chat event
- [x] Add message event

### PR DESCRIPTION
**Pull request by ptarho**
I decided to create socket as a plugin in order to get access to the server instance

### CODE REVIEW

File: backend/src/common/plugins/socket/socket.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    done,
+) => {
+    const io = new Server(fastify.server, {
+        cors: {
+            origin: '*',
+            credentials: true,
+        },
+    });
+
+    io.on(SocketEvent.CONNECTION, (socket) => {
```
Comment (nika-sirak) [inline backend/src/common/plugins/socket/socket.plugin.ts]: io.of(SocketNamespace.CHAT).on(SocketEvent.CONNECTION, (socket) => {

**Review by MaksymGavatuk — APPROVED** [review]
Looks good as for me. Only Nika comment

File: backend/src/common/plugins/socket/socket.plugin.ts (hunk 1 above)
Comment (ptarho) [reply backend/src/common/plugins/socket/socket.plugin.ts]: Fixed

**Review by anton-otroshchenko — APPROVED** [review]

File: backend/src/common/plugins/socket/socket.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        socket.on(SocketEvent.CHAT_LEAVE_ROOM, (chatId: string) => {
+            return socket.leave(chatId);
+        });
+
+        socket.on(
+            SocketEvent.CHAT_CREATE_MESSAGE,
+            (payload: ChatMessageGetAllItemResponseDto) => {
+                if (socket.rooms.has(payload.chatId)) {
+                    socket
+                        .to(payload.chatId)
```
Comment (nika-sirak) [inline backend/src/common/plugins/socket/socket.plugin.ts]: or .to(payload.reciever.id)?

File: backend/src/common/plugins/socket/socket.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            },
+        );
+
+        socket.on('disconnect', () => {
+            const userId = [...connectedUsers.values()].find(
+                (it) => it === socket.id,
+            );
+            connectedUsers.delete(userId);
+        });
+    });
```
Comment (nikita-remeslov) [inline backend/src/common/plugins/socket/socket.plugin.ts]: create private method `chatHandler` and pass it to `on`.
Also, it should be Class Socket

File: backend/src/common/plugins/socket/socket.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                const receiverId = connectedUsers.get(payload.receiver.id);
+                if (receiverId) {
+                    socket
+                        .to(receiverId)
+                        .emit(SocketEvent.CHAT_ADD_MESSAGE, payload);
+                }
+            },
+        );
+
+        socket.on('disconnect', () => {
```
Comment (nikita-remeslov) [inline backend/src/common/plugins/socket/socket.plugin.ts]: should be in the enum

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/common/plugins/socket/socket.plugin.ts (hunk 2 above)
Comment (ptarho) [reply backend/src/common/plugins/socket/socket.plugin.ts]: Done

File: backend/src/common/plugins/socket/socket.plugin.ts (hunk 4 above)
Comment (ptarho) [reply backend/src/common/plugins/socket/socket.plugin.ts]: fixed

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**